### ISSUE TITLE
clang14 hangs forever cross-compiling libraw/crx.cpp for riscv64/riscv32

### ISSUE THREAD

**Issue by kraj** [llvm/llvm-project#53831]
Attached testcase hangs the compiler when using -O2 or -Os, it works ok with -O1

```

clang version 14.0.0 (https://github.com/llvm/llvm-project d6f290ec136b7faef956ee2b9bfc0fe0ee0f85ab)
Target: riscv32-yoe-linux
Thread model: posix
InstalledDir: /mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/libraw/0.20.2-r0/recipe-sysroot-native/usr/bin/riscv32-yoe-linux
Found candidate GCC installation: /usr/lib/gcc/riscv64-linux-gnu/11.1.0
Found candidate GCC installation: /usr/lib64/gcc/riscv64-linux-gnu/11.1.0
Selected GCC installation: /usr/lib/gcc/riscv64-linux-gnu/11.1.0
Candidate multilib: .;@m32@mabi=ilp32d
Candidate multilib: .;@m64@mabi=lp64d
Selected multilib: .;@m32@mabi=ilp32d
clang-14: warning: treating 'cpp-output' input as 'c++-cpp-output' when in C++ mode, this behavior is deprecated [-Wdeprecated]
 "/mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/libraw/0.20.2-r0/recipe-sysroot-native/usr/bin/clang-14" -cc1 -triple riscv32-yoe-linux -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name crx.i -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -target-feature +m -target-feature +a -target-feature +f -target-feature +d -target-feature +c -target-feature +relax -target-feature -save-restore -target-abi ilp32d -msmall-data-limit 8 -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -v -fcoverage-compilation-dir=/mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/libraw/0.20.2-r0/build -resource-dir /mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/libraw/0.20.2-r0/recipe-sysroot-native/usr/lib/clang/14.0.0 -O2 -fdeprecated-macro -fdebug-compilation-dir=/mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/libraw/0.20.2-r0/build -ferror-limit 19 -stack-protector 2 -fno-signed-char -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/crx-25be68.o -x c++-cpp-output /tmp/crx.i
clang -cc1 version 14.0.0 based upon LLVM 14.0.0 default target x86_64-unknown-linux-gnu
#include "..." search starts here:
End of search list.

```
[crx.tar.gz](https://github.com/llvm/llvm-project/files/8063695/crx.tar.gz)

**Comment by asb** [comment]
Thank you, I can reproduce locally and the issue didn't exist in Clang 13. Looking at `-mllvm -debug` the loop seems to be:

```
ombining: t58213: i32 = mul t155, Constant:i32<26215>
Creating new node: t58215: i32 = mul t190, Constant:i32<26215>
Creating new node: t58216: i32 = add t58215, Constant:i32<26215>
 ... into: t58216: i32 = add t58215, Constant:i32<26215>

Combining: t155: i32 = add t190, Constant:i32<1>

Combining: t188: i32 = sra t58216, Constant:i32<17>

Combining: t148: i32 = srl t58216, Constant:i32<31>

Combining: t58216: i32 = add t58215, Constant:i32<26215>
Creating new node: t58217: i32 = mul t155, Constant:i32<26215>
Creating constant: t58218: i32 = Constant<0>
 ... into: t58217: i32 = mul t155, Constant:i32<26215>

Combining: t190: i32,ch = load<(load (s8) from %ir.fltTapH, align 4, !tbaa !81), sext from i8> t0, t16, undef:i32

Combining: t148: i32 = srl t58217, Constant:i32<31>

Combining: t188: i32 = sra t58217, Constant:i32<17>

Combining: t58217: i32 = mul t155, Constant:i32<26215>
Creating new node: t58219: i32 = mul t190, Constant:i32<26215>
Creating new node: t58220: i32 = add t58219, Constant:i32<26215>
 ... into: t58220: i32 = add t58219, Constant:i32<26215>

Combining: t155: i32 = add t190, Constant:i32<1>

Combining: t188: i32 = sra t58220, Constant:i32<17>

Combining: t148: i32 = srl t58220, Constant:i32<31>

Combining: t58220: i32 = add t58219, Constant:i32<26215>
Creating new node: t58221: i32 = mul t155, Constant:i32<26215>
Creating constant: t58222: i32 = Constant<0>
 ... into: t58221: i32 = mul t155, Constant:i32<26215>
........(continues)
```

**Comment by asb** [comment]
The culprit is transformAddImmMulImm. @benshi001 - could you please take a look?

Here is a bugpoint-reduced testcase (`llc -mtriple=riscv32 -mattr=+m` will do the trick):
```
%struct.CrxWaveletTransform.0.1.2.3.4.39.40 = type { i32*, i32*, i32*, i32*, [8 x i32*], i16, i16, i8, i16, i16 }

define dso_local void @_Z24crxIdwt53FilterTransformP12CrxPlaneCompj.if.end23.split(%struct.CrxWaveletTransform.0.1.2.3.4.39.40* %0, i32 %level, i32** %.out2, i32** %.out3, i32** %a) {
newFuncRoot:
  %1 = load i8, i8* undef, align 4
  %2 = sext i8 %1 to i16
  %rem.lhs.trunc = add nsw i16 %2, 1
  %rem848 = srem i16 %rem.lhs.trunc, 5
  %narrow = add nsw i16 %rem848, 3
  %add31 = sext i16 %narrow to i32
  %arrayidx32 = getelementptr inbounds %struct.CrxWaveletTransform.0.1.2.3.4.39.40, %struct.CrxWaveletTransform.0.1.2.3.4.39.40* %0, i32 %level, i32 4, i32 %add31
  %3 = load i32*, i32** %arrayidx32, align 4
  store i32* %3, i32** %.out2, align 4
  %rem37.lhs.trunc = add nsw i16 %2, 2
  %rem37849 = srem i16 %rem37.lhs.trunc, 5
  %narrow850 = add nsw i16 %rem37849, 3
  %add38 = sext i16 %narrow850 to i32
  %arrayidx39 = getelementptr inbounds %struct.CrxWaveletTransform.0.1.2.3.4.39.40, %struct.CrxWaveletTransform.0.1.2.3.4.39.40* %0, i32 %level, i32 4, i32 %add38
  %4 = load i32*, i32** %arrayidx39, align 4
  store i32* %4, i32** %.out3, align 4
  unreachable
}
```

**Comment by kraj** [comment]
I have another case seen with riscv32 backend when using `-O2` which hangs in similar fashion. I wonder if it is related as well. I am attaching the test case.
[edc_ecc.i.tar.gz](https://github.com/llvm/llvm-project/files/8091997/edc_ecc.i.tar.gz)

**Comment by kraj** [comment]
another hanging testcase on rv64

[libntfs_3g_la-runlist.i.tar.gz](https://github.com/llvm/llvm-project/files/8092502/libntfs_3g_la-runlist.i.tar.gz)

**Comment by asb** [comment]
I've stepped through the first test case in a little more detail:
* The issue is a loop between the target-independent transform in DAGCombiner::visitMUL triggered by isMulAddWtihConstProfitable and one of the transformations in RISCVISelLowering::transformAddImmMulImm.
* The target-independent transform is doing `(mul (add x, c1), c2) -> (add (mul x, c2), c1*c2)`.
* The target-dependent transform is doing `(add (mul x, c0), c1) -> (add (mul (add x, c1/c0), c0), c1%c0)`
* RISCVISelLowering::transformAdImmMulImm attempts to exit early if it believes the transformation will be undone (if `c0*(c1/c0)` is simm12). This is a property that would be tested in RISCVISelLowering;:isMulAddWithConstProfitable. However, this condition isn't true, but there's still a loop.
* The logic error is that DAGCombiner::isMulAddWithConstProfitable might still return true (and potentially undo the target-specific transformation) if TI.isMulAddWithConstProfitable returns false. In the test case in this bug report, the condition labeled "check to see if a future expansion will give us a common multiply" is true.

@benshi001, @topperc, any thoughts on a preferred fix? Making transformAddImmMulImm bail out early if `!N0C->hasOneUse()` would I think conservatively avoid any transformation that the current `DAGCombiner::isMulAddWithConstProfitable` would make.

Arguably, the behaviour of `TargetLowering::isMulAddWithConstProfitable` returning false should be improved as well, but I suspect the result would be that the second part of DAGCombiner::isMulAddWithConstProfitable is never executed for targets that implement the hook.

**Comment by asb** [comment]
I've submitted https://reviews.llvm.org/D120332 which implements a minimally invasive fix.

**Comment by asb** [comment]
Thanks for reporting this and providing several test cases. A fix has now been committed in to `main` branch, and I've requested a backport to the 14.x branch. I'll leave this issue open until that backport request is merged.

**Comment by kraj** [comment]
Thanks a bunch @asb I am trying it out on 14.x as well I willl update if there are any issues left

**Comment by kraj** [comment]
@asb I have tested the newly build clang and all cases where it was hanging on RV64 seems to be fixed. I will check RV32 soon

**Comment by tstellar** [comment]
Merged: ee0ae47691d3
